# Logistic regression models

In [1]:
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Load the MNIST dataset
mnist = fetch_openml('mnist_784')

/home/dgist/.local/lib/python3.10/site-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [2]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=0.2, random_state=42)

# Initialize variables to store the best model and its accuracy
best_model = None
best_accuracy = 0

# Try different values for regularization parameter C
C_values = [0.001, 0.01, 0.1, 1, 10, 100, 1000]

In [3]:
for C in C_values:
    # Train logistic regression model
    model = LogisticRegression(C=C, max_iter=1000)
    model.fit(X_train, y_train)
    
    # Predict on the test set
    y_pred = model.predict(X_test)
    
    # Calculate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    
    # Check if current model is better than previous best model
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_model = model

/home/dgist/.local/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/dgist/.local/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

In [4]:
print("Best model accuracy:", best_accuracy)
print("Best model parameters:", best_model.get_params())

Best model accuracy: 0.9167142857142857
Best model parameters: {'C': 0.001, 'class_weight': None, 'dual': False, 'fit_intercept': True, 'intercept_scaling': 1, 'l1_ratio': None, 'max_iter': 1000, 'multi_class': 'auto', 'n_jobs': None, 'penalty': 'l2', 'random_state': None, 'solver': 'lbfgs', 'tol': 0.0001, 'verbose': 0, 'warm_start': False}


### Results
Best model accuracy: 0.9167142857142857  
Best model parameters: {'C': 0.001, 'class_weight': None, 'dual': False, 'fit_intercept': True, 'intercept_scaling': 1, 'l1_ratio': None, 'max_iter': 1000, 'multi_class': 'auto', 'n_jobs': None, 'penalty': 'l2', 'random_state': None, 'solver': 'lbfgs', 'tol': 0.0001, 'verbose': 0, 'warm_start': False}

# K-NN Classifiers

In [8]:
from sklearn.neighbors import KNeighborsClassifier


# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=0.2, random_state=42)

# Initialize variables to store the best model and its accuracy
best_model = None
best_accuracy = 0

# Try different values for K
k_values = [3, 5, 7, 9, 11]

In [9]:
for k in k_values:
    # Train K-NN classifier
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    
    # Predict on the test set
    y_pred = knn.predict(X_test)
    
    # Calculate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    
    # Check if current model is better than previous best model
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_model = knn

In [10]:
print("Best model accuracy:", best_accuracy)
print("Best model parameters:", best_model.get_params())

Best model accuracy: 0.9712857142857143
Best model parameters: {'algorithm': 'auto', 'leaf_size': 30, 'metric': 'minkowski', 'metric_params': None, 'n_jobs': None, 'n_neighbors': 3, 'p': 2, 'weights': 'uniform'}


### Result
Best model accuracy: 0.9712857142857143  
Best model parameters: {'algorithm': 'auto', 'leaf_size': 30, 'metric': 'minkowski', 'metric_params': None, 'n_jobs': None, 'n_neighbors': 3, 'p': 2, 'weights': 'uniform'}

# SVM Classifiers

In [3]:
from sklearn.svm import SVC

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=0.2, random_state=42)

# Initialize variables to store the best model and its accuracy
best_model = None
best_accuracy = 0

# Try different values for the regularization parameter C and kernel function
C_values = [0.1, 1, 10]
kernel_values = ['linear', 'rbf']

In [4]:
for C in C_values:
    for kernel in kernel_values:
        # Train SVM classifier
        svm = SVC(C=C, kernel=kernel)
        svm.fit(X_train, y_train)
        
        # Predict on the test set
        y_pred = svm.predict(X_test)
        
        # Calculate accuracy
        accuracy = accuracy_score(y_test, y_pred)
        
        # Check if current model is better than previous best model
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_model = svm

In [ ]:
print("Best model accuracy:", best_accuracy)
print("Best model parameters:", best_model.get_params())

# Random forest classifiers

In [8]:
from sklearn.ensemble import RandomForestClassifier

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=0.2, random_state=42)

# Initialize variables to store the best model and its accuracy
best_model = None
best_accuracy = 0

# Try different values for the number of trees in the forest
n_estimators_values = [50, 100, 200]

In [9]:
for n_estimators in n_estimators_values:
    # Train Random Forest classifier
    rf = RandomForestClassifier(n_estimators=n_estimators, random_state=42)
    rf.fit(X_train, y_train)
    
    # Predict on the test set
    y_pred = rf.predict(X_test)
    
    # Calculate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    
    # Check if current model is better than previous best model
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_model = rf

In [10]:
print("Best model accuracy:", best_accuracy)
print("Best model parameters:", best_model.get_params())

Best model accuracy: 0.9681428571428572
Best model parameters: {'bootstrap': True, 'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': 'sqrt', 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 200, 'n_jobs': None, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}


### Results
Best model accuracy: 0.9681428571428572  
Best model parameters: {'bootstrap': True, 'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': 'sqrt', 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 200, 'n_jobs': None, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}

# Gradient Boosting classifiers

In [4]:
from sklearn.ensemble import GradientBoostingClassifier

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=0.2, random_state=42)

# Initialize variables to store the best model and its accuracy
best_model = None
best_accuracy = 0

# Try different values for the number of trees in the ensemble
# n_estimators_values = [50, 100, 200]
n_estimators_values = [50]

In [5]:
for n_estimators in n_estimators_values:
    # Train Gradient Boosting classifier
    gb = GradientBoostingClassifier(n_estimators=n_estimators, random_state=42)
    gb.fit(X_train, y_train)
    
    # Predict on the test set
    y_pred = gb.predict(X_test)
    
    # Calculate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    
    # Check if current model is better than previous best model
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_model = gb

KeyboardInterrupt: 

In [ ]:
print("Best model accuracy:", best_accuracy)
print("Best model parameters:", best_model.get_params())